# CAMELYON17 Dataset Audit

This notebook performs a reproducible structural and integrity audit of the
CAMELYON17 dataset used in the robustness study.

The audit covers:

- dataset size and structure
- official WILDS split assignment
- center and tumor-label distributions
- center–tumor association
- patient and slide structure
- metadata consistency
- image–metadata integrity

This notebook does not modify the official dataset splits or labels.

In [10]:
from getpass import getpass

github_token = getpass("GitHub token: ")

GitHub token: ··········


In [11]:
import subprocess

repo_url = "https://github.com/s-sana-sharifi/CAMELYON17-robustness.git"
repo_dir = "/content/CAMELYON17-robustness"

result = subprocess.run(
    [
        "git", "clone",
        f"https://s-sana-sharifi:{github_token}@github.com/"
        "s-sana-sharifi/CAMELYON17-robustness.git",
        repo_dir,
    ],
    capture_output=True,
    text=True,
)

print(result.stdout)
print(result.stderr)

if result.returncode == 0:
    print("Clone successful.")


Cloning into '/content/CAMELYON17-robustness'...

Clone successful.


In [12]:
from pathlib import Path
import sys
import os

REPO_ROOT = Path("/content/CAMELYON17-robustness")

sys.path.insert(0, str(REPO_ROOT))
os.chdir(REPO_ROOT)

print("Repository:", Path.cwd())
print("src exists:", (REPO_ROOT / "src").exists())
print("audit.py exists:", (REPO_ROOT / "src/data/audit.py").exists())
print("splits.py exists:", (REPO_ROOT / "src/data/splits.py").exists())

Repository: /content/CAMELYON17-robustness
src exists: True
audit.py exists: True
splits.py exists: True


In [13]:
from src.data.audit import (
    load_metadata,
    summarize_dataset,
    summarize_splits,
    summarize_centers,
    summarize_labels,
    center_label_distribution,
    tumor_probability_by_center,
    patient_structure,
    slide_structure,
    check_metadata_consistency,
    summarize_patient_center_relationship,
    summarize_slide_center_relationship,
)

from src.data.splits import assign_official_split_ids

print("Imports successful.")

Imports successful.


In [15]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [16]:
from pathlib import Path

ARCHIVE = Path(
    "/content/drive/MyDrive/"
    "CAMELYON17-robustness/dataset/camelyon17_v1.0.tar"
)

print("Archive exists:", ARCHIVE.exists())

if ARCHIVE.exists():
    print(f"Archive size: {ARCHIVE.stat().st_size / (1024**3):.2f} GB")

Archive exists: True
Archive size: 10.68 GB


In [18]:
from pathlib import Path
import tarfile

ARCHIVE = Path(
    "/content/drive/MyDrive/"
    "CAMELYON17-robustness/dataset/camelyon17_v1.0.tar"
)

EXTRACT_ROOT = Path("/content/wilds_data")

EXTRACT_ROOT.mkdir(parents=True, exist_ok=True)

print("Archive:", ARCHIVE)
print(f"Archive size: {ARCHIVE.stat().st_size / (1024**3):.2f} GB")
print("Extracting...")

with tarfile.open(ARCHIVE, "r") as tar:
    tar.extractall(EXTRACT_ROOT)

print("Extraction complete.")

Archive: /content/drive/MyDrive/CAMELYON17-robustness/dataset/camelyon17_v1.0.tar
Archive size: 10.68 GB
Extracting...


/tmp/ipykernel_1338/3827924951.py:18: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tar.extractall(EXTRACT_ROOT)


Extraction complete.


In [19]:
from pathlib import Path

matches = list(Path("/content/wilds_data").rglob("metadata.csv"))

print(f"Found {len(matches)} metadata.csv file(s):")

for path in matches:
    print(path)

Found 1 metadata.csv file(s):
/content/wilds_data/camelyon17_v1.0/metadata.csv


In [20]:
METADATA_PATH = matches[0]
DATASET_ROOT = METADATA_PATH.parent
PATCHES_ROOT = DATASET_ROOT / "patches"

print("Dataset root:", DATASET_ROOT)
print("Metadata:", METADATA_PATH)
print("Patches:", PATCHES_ROOT)

print("\nMetadata exists:", METADATA_PATH.exists())
print("Patches exists:", PATCHES_ROOT.exists())

Dataset root: /content/wilds_data/camelyon17_v1.0
Metadata: /content/wilds_data/camelyon17_v1.0/metadata.csv
Patches: /content/wilds_data/camelyon17_v1.0/patches

Metadata exists: True
Patches exists: True


In [21]:
metadata = load_metadata(METADATA_PATH)

print(f"Loaded metadata with {len(metadata):,} rows.")

print("\nColumns:")
print(metadata.columns.tolist())

Loaded metadata with 455,954 rows.

Columns:
['Unnamed: 0', 'patient', 'node', 'x_coord', 'y_coord', 'tumor', 'slide', 'center', 'split']


In [23]:
dataset_summary = summarize_dataset(metadata)

for key, value in dataset_summary.items():
    print(f"{key}: {value:,}")

num_samples: 455,954
num_patients: 43
num_slides: 50
num_centers: 5


In [24]:
official_split = assign_official_split_ids(
    metadata["center"],
    metadata["split"],
)

metadata_audit = metadata.copy()
metadata_audit["official_split"] = official_split

In [25]:
OFFICIAL_SPLIT_NAMES = {
    0: "Train",
    1: "Validation (ID)",
    2: "Test (OOD)",
    3: "Validation (OOD)",
}

split_summary = (
    metadata_audit["official_split"]
    .value_counts()
    .sort_index()
    .rename_axis("split_id")
    .reset_index(name="num_samples")
)

split_summary["split"] = split_summary["split_id"].map(OFFICIAL_SPLIT_NAMES)
split_summary["proportion"] = (
    split_summary["num_samples"] / len(metadata_audit)
)

print(split_summary[
    ["split_id", "split", "num_samples", "proportion"]
].to_string(index=False))

 split_id            split  num_samples  proportion
        0            Train       302436    0.663304
        1  Validation (ID)        33560    0.073604
        2       Test (OOD)        85054    0.186541
        3 Validation (OOD)        34904    0.076552


In [26]:
center_summary = summarize_centers(metadata)

print(center_summary.to_string(index=False))

 center  num_samples  proportion
      0        59436    0.130355
      1        34904    0.076552
      2        85054    0.186541
      3       129838    0.284761
      4       146722    0.321791


In [27]:
label_summary = summarize_labels(metadata)

print(label_summary.to_string(index=False))

 tumor  num_samples  proportion
     0       227977         0.5
     1       227977         0.5


In [28]:
center_tumor = center_label_distribution(metadata)

print(center_tumor.to_string(index=False))

 center  tumor  num_samples  proportion_within_center
      0      0        29718                       0.5
      0      1        29718                       0.5
      1      0        17452                       0.5
      1      1        17452                       0.5
      2      0        42527                       0.5
      2      1        42527                       0.5
      3      0        64919                       0.5
      3      1        64919                       0.5
      4      0        73361                       0.5
      4      1        73361                       0.5


In [29]:
tumor_prob_by_center = tumor_probability_by_center(metadata)

print(tumor_prob_by_center.to_string(index=False))

 center  p_tumor_1
      0        0.5
      1        0.5
      2        0.5
      3        0.5
      4        0.5


In [30]:
patient_summary = patient_structure(metadata)

print(patient_summary.to_string(index=False))

 center  patient  num_samples  num_slides
      0        4         4316           1
      0        9         4597           1
      0       10         4046           1
      0       12        13637           1
      0       15        10554           2
      0       16         3427           1
      0       17        18859           3
      1       20         7449           2
      1       21         1705           1
      1       22        10404           1
      1       24         3854           2
      1       34         4971           1
      1       36         2659           1
      1       38         1275           1
      1       39         2587           1
      2       40         3810           1
      2       41         3694           1
      2       42         7210           1
      2       44         5288           1
      2       45         7727           1
      2       46         8149           2
      2       48         4556           1
      2       51        31878     

In [31]:
slide_summary = slide_structure(metadata)

print(slide_summary.to_string(index=False))

 center  patient  slide  num_samples  num_patients
      0        4      0         4316             1
      0        9      1         4597             1
      0       10      2         4046             1
      0       12      3        13637             1
      0       15      4         7294             1
      0       15      5         3260             1
      0       16      6         3427             1
      0       17      7         1596             1
      0       17      8        13455             1
      0       17      9         3808             1
      1       20     10         2078             1
      1       20     11         5371             1
      1       21     12         1705             1
      1       22     13        10404             1
      1       24     14          888             1
      1       24     15         2966             1
      1       34     16         4971             1
      1       36     17         2659             1
      1       38     18        

In [32]:
consistency_issues = check_metadata_consistency(metadata)

print("Consistency issues:")

if consistency_issues:
    for key, value in consistency_issues.items():
        print(f"- {key}: {value}")
else:
    print("No consistency issues found.")

Consistency issues:
No consistency issues found.


In [33]:
png_files = list(PATCHES_ROOT.rglob("*.png"))

print(f"PNG files found: {len(png_files):,}")
print(f"Metadata rows: {len(metadata):,}")

PNG files found: 455,954
Metadata rows: 455,954


In [34]:
for path in png_files[:5]:
    print(path)

/content/wilds_data/camelyon17_v1.0/patches/patient_046_node_3/patch_patient_046_node_3_x_19712_y_12512.png
/content/wilds_data/camelyon17_v1.0/patches/patient_046_node_3/patch_patient_046_node_3_x_8672_y_3776.png
/content/wilds_data/camelyon17_v1.0/patches/patient_046_node_3/patch_patient_046_node_3_x_7136_y_4192.png
/content/wilds_data/camelyon17_v1.0/patches/patient_046_node_3/patch_patient_046_node_3_x_13600_y_3552.png
/content/wilds_data/camelyon17_v1.0/patches/patient_046_node_3/patch_patient_046_node_3_x_18048_y_11008.png


In [35]:
import re

IMAGE_PATTERN = re.compile(
    r"patch_patient_(\d+)_node_(\d+)_x_(\d+)_y_(\d+)\.png$"
)

image_identities = []

for path in png_files:
    match = IMAGE_PATTERN.search(path.name)

    if match is None:
        raise ValueError(f"Unexpected filename format: {path}")

    patient, node, x_coord, y_coord = map(int, match.groups())

    image_identities.append(
        (patient, node, x_coord, y_coord)
    )

image_df = pd.DataFrame(
    image_identities,
    columns=["patient", "node", "x_coord", "y_coord"],
)

print(f"Parsed image identities: {len(image_df):,}")
print(f"Unique image identities: {len(image_df.drop_duplicates()):,}")

Parsed image identities: 455,954
Unique image identities: 455,954


In [36]:
metadata_identities = metadata[
    ["patient", "node", "x_coord", "y_coord"]
].copy()

print(
    f"Unique metadata identities: "
    f"{len(metadata_identities.drop_duplicates()):,}"
)

Unique metadata identities: 455,954


In [37]:
image_set = set(map(tuple, image_df.to_numpy()))
metadata_set = set(map(tuple, metadata_identities.to_numpy()))

images_missing_from_metadata = image_set - metadata_set
metadata_missing_from_images = metadata_set - image_set

print(
    "Images missing from metadata:",
    len(images_missing_from_metadata),
)

print(
    "Metadata entries missing from images:",
    len(metadata_missing_from_images),
)

print(
    "Exact identity match:",
    image_set == metadata_set,
)

Images missing from metadata: 0
Metadata entries missing from images: 0
Exact identity match: True


1. CAMELYON17 contains 455,954 image patches from 43 patients, 50 slides, and 5 centers.
2. The official WILDS split reconstruction gives 302,436 training samples, 33,560 ID validation samples, 34,904 OOD validation samples, and 85,054 OOD test samples.
3. Tumor labels are exactly balanced overall and within every center.
4. There is no aggregate center–tumor association in the metadata: P(tumor=1 | center) = 0.5 for every center.
5. Each patient belongs to one center, and each slide belongs to one patient and one center.
6. All 455,954 image patches have an exact metadata identity match.
